src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs

```diff
@@ -230,40 +236,42 @@ public override async ValueTask<FormulaValue> Visit(CallNode node, EvalVisitorCo
-            {
-                var result = await asyncFunc.InvokeAsync(args, _cancellationToken);
-                return result;
-            }
-            else if (func is UserDefinedTexlFunction udtf)
-            {
-                // $$$ Should add _runtimeConfig
-                var result = await udtf.InvokeAsync(args, _cancellationToken, context.StackDepthCounter.Increment());
-                return result;
-            }
-            else if (func is CustomTexlFunction customTexlFunc)
-            {
-                var result = await customTexlFunc.InvokeAsync(FunctionServices, args, _cancellationToken);
-                return result;
-            }
-            else
-            {
-                if (FunctionImplementations.TryGetValue(func, out var ptr))
-                {
-                    var result = await ptr(this, context.IncrementStackDepthCounter(childContext), node.IRContext, args);
-
-                    if (IfFunction.CanCheckIfReturn(func))
-                    {
-                        Contract.Assert(result.IRContext.ResultType == node.IRContext.ResultType || result is ErrorValue || result.IRContext.ResultType is BlankType);
-                    }
-
-                    return result;
-                }
-
-                return CommonErrors.NotYetImplementedError(node.IRContext, $"Missing func: {func.Name}");
-            }
+            var childContext = context.SymbolContext.WithScope(node.Scope);
+
+            FormulaValue result;
+            if (func is IAsyncTexlFunction asyncFunc)
+            {
+                result = await asyncFunc.InvokeAsync(args, _cancellationToken);
+            }
+            else if (func is UserDefinedTexlFunction udtf)
+            {
+                // $$$ Should add _runtimeConfig
+                result = await udtf.InvokeAsync(args, _cancellationToken, context.StackDepthCounter.Increment());
+            }
+            else if (func is CustomTexlFunction customTexlFunc)
+            {
+                // If custom function throws an exception, don't catch it - let it propagate up to the host.
+                result = await customTexlFunc.InvokeAsync(FunctionServices, args, _cancellationToken);
+            }
+            else
+            {
+                if (FunctionImplementations.TryGetValue(func, out var ptr))
+                {
+                    result = await ptr(this, context.IncrementStackDepthCounter(childContext), node.IRContext, args);
+
+                    if (IfFunction.CanCheckIfReturn(func))
+                    {
+                        Contract.Assert(result.IRContext.ResultType == node.IRContext.ResultType || result is ErrorValue || result.IRContext.ResultType is BlankType);
+                    }
+                }
+                else
+                {
+                    result = CommonErrors.NotYetImplementedError(node.IRContext, $"Missing func: {func.Name}");
+                }
+            }
+
+            CheckCancel();

```

Comment:
Why add here just before returning the result?

Reply:
We're returning from this function, but we're still in the middle of an execution. 
This also checks for memory usage - so checking after a function called keeps us close to the original point of failure and a small window.

Reply:
The name of this function isn't related to what it's really doing then
+ Shouldn't memory allocations be checked before allocating?
+ Checking for cancellation token at this point is not the right place.

Reply:
We check both before and after. 
If a function managed to allocate and go over the limit, this is an opportunity to catch it immediately.